src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java

```diff
@@ -166,15 +179,99 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+            } catch (IOException e) {
+                LOGGER.log(Level.SEVERE, "cannot create knowhosts file", e);
+            }
+        }
+        return new SshdSessionFactory() {
+            private Path tmpKey;
+
+            @Override
+            protected SshConfigStore createSshConfigStore(File homeDir, File configFile, String localUserName) {
+                return new OpenSshConfigFile(homeDir, configFile, localUserName) {
+                    @Override
+                    public HostEntry lookup(String hostName, int port, String userName) {
+                        HostEntry hostEntry = super.lookup(hostName, port, userName);
+                        return abstractJGitHostKeyVerifier.customizeHostEntry(hostEntry);
+                    }
+                };
+            }
+
+            @Override
+            public File getHomeDirectory() {
+                return SystemUtils.getUserHome();
+            }
+
+            @Override
+            public File getSshDirectory() {
+                return SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE.getParentFile();
+            }
+
+            @Override
+            public SshdSession getSession(URIish uri, CredentialsProvider credentialsProvider, FS fs, int tms)
+                    throws TransportException {
+                SshdSession sshdSession = super.getSession(uri, credentialsProvider, fs, tms);
+                sshdSession.addCloseListener(sshdSession1 -> {
+                    if (tmpKey != null) {
+                        try {
+                            Files.deleteIfExists(tmpKey);
+                        } catch (IOException e) {
+                            if (LOGGER.isLoggable(Level.FINE))
+                                LOGGER.log(Level.FINE, "ignore fail to delete tmp key", e);
+                        }
+                    }
+                });
+                return sshdSession;
+            }
+
+            @Override
+            protected List<Path> getDefaultIdentities(File sshDir) {
+                SSHUserPrivateKey sshUserPrivateKey = credentialsProvider.getCredentials().values().stream()
+                        .filter(standardCredentials -> standardCredentials instanceof SSHUserPrivateKey)
+                        .map(SSHUserPrivateKey.class::cast)
+                        .findFirst()
+                        .orElse(null);
+                if (sshUserPrivateKey != null) {
+                    try {
+                        // be sure parent directories are here
+                        Files.createDirectories(sshDir.getAbsoluteFile().toPath());
+                        tmpKey = Files.createTempFile(sshDir.getAbsoluteFile().toPath(), "key", ".priv");
+                        tmpKey.toFile().deleteOnExit();
+                        Files.write(tmpKey, sshUserPrivateKey.getPrivateKeys());
+                        return Collections.singletonList(tmpKey);
```

Comment:
I think should be using a `workspace@tmp` or the like?  
Files.createDirectories will in any case, create the `.ssh` directory with incorrect perms which can/will break any use of the native ssh on the host.